<?xml version="1.0"?>
<!-- created with Phobos 0.7 -->
  <robot name="diff_robot" xmlns:xacro="http://www.ros.org/wiki/xacro" >

    <xacro:property name="wheel_friction" value="1.0" />

    <xacro:include filename="$(find diff_robot_gazebo)/urdf/lidar.urdf.xacro"/>
    <xacro:lidar name="laser_front" parent_frame="base_link" x="0.45" y="0.0" z="0.0" roll="0.0" pitch="0.0" yaw="0.0" />

    <gazebo>
      <!-- Simulated vehicle interface -->
      <plugin name="diff_robot_interface_plugin" filename="libdiff_interface_plugin.so" >
        <ground_truth>$(arg ground_truth)</ground_truth>
      </plugin>

      <!-- Publish current steering angles -->
      <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
          <robotNamespace>$(arg robot_name)</robotNamespace>
          <jointName>
              left_wheel_joint, right_wheel_joint
          </jointName>
          <updateRate>100</updateRate>
          <alwaysOn>true</alwaysOn>
      </plugin>
    </gazebo>

    <link name="base_footprint">
    </link>

    <link name="base_link">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.1"/>
        <inertia ixx="0.16" ixy="0" ixz="0" iyy="0.16" iyz="0" izz="0.24"/>
      </inertial>
      <visual name="Cylinder.007">
        <origin xyz="-0.32 0 -0.12" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.011.dae" scale="0.05 0.05 0.2"/>
        </geometry>
        <material name="Material.011"/>
      </visual>
      <visual name="Cylinder.008">
        <origin xyz="0.32 0 -0.12" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.012.dae" scale="0.05 0.05 0.2"/>
        </geometry>
        <material name="Material.012"/>
      </visual>
      <visual name="Cylinder.009">
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.013.dae" scale="0.4 0.4 0.2"/>
        </geometry>
        <material name="Material.015"/>
      </visual>
      <visual name="Sphere">
        <origin xyz="0.32 0 -0.35" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Sphere.dae" scale="0.05 0.05 0.05"/>
        </geometry>
        <material name="Material.013"/>
      </visual>
      <visual name="Sphere.001">
        <origin xyz="-0.32 0 -0.35" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Sphere.001.dae" scale="0.05 0.05 0.05"/>
        </geometry>
        <material name="Material.014"/>
      </visual>
      <collision name="Sphere.002">
        <origin xyz="0.32 0 -0.35" rpy="0 0 0"/>
        <geometry>
          <sphere radius="0.05"/>
        </geometry>
      </collision>
      <collision name="Sphere.003">
        <origin xyz="-0.32 0 -0.35" rpy="0 0 0"/>
        <geometry>
          <sphere radius="0.05"/>
        </geometry>
      </collision>
      <collision name="chassis_collision">
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="0.4" length="0.4"/>
        </geometry>
      </collision>
    </link>

    <link name="left_wheel">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="20.0"/>
        <inertia ixx="0.010833" ixy="0" ixz="0" iyy="0.010833" iyz="0" izz="0.02"/>
      </inertial>
      <visual name="Cylinder.004">
        <origin xyz="0 0 0.05" rpy="1.570796 0 1.570796"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.008.dae" scale="0.02 0.02 0.15"/>
        </geometry>
        <material name="Material.008"/>
      </visual>
      <visual name="Cylinder.005">
        <origin xyz="0 0 0" rpy="-3.141593 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.009.dae" scale="0.2 0.2 0.05"/>
        </geometry>
        <material name="Material.009"/>
      </visual>
      <visual name="Cylinder.006">
        <origin xyz="0 0 0.05" rpy="1.570796 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.010.dae" scale="0.02 0.02 0.15"/>
        </geometry>
        <material name="Material.010"/>
      </visual>
      <collision name="left_wheel_collision">
        <origin xyz="0 0 0" rpy="-3.141593 0 0"/>
        <geometry>
          <cylinder radius="0.2" length="0.1"/>
        </geometry>
      </collision>
    </link>

    <link name="right_wheel">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="20.0"/>
        <inertia ixx="0.010833" ixy="0" ixz="0" iyy="0.010833" iyz="0" izz="0.02"/>
      </inertial>
      <visual name="Cylinder.001">
        <origin xyz="0 0 -0.05" rpy="1.570796 0 1.570796"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.003.dae" scale="0.02 0.02 0.15"/>
        </geometry>
        <material name="Material.003"/>
      </visual>
      <visual name="Cylinder.002">
        <origin xyz="0 0 0" rpy="-3.141593 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.002.dae" scale="0.2 0.2 0.05"/>
        </geometry>
        <material name="Material.002"/>
      </visual>
      <visual name="Cylinder.003">
        <origin xyz="0 0 -0.05" rpy="1.570796 0 0"/>
        <geometry>
          <mesh filename="package://diff_robot_gazebo/meshes/Cylinder.004.dae" scale="0.02 0.02 0.15"/>
        </geometry>
        <material name="Material.004"/>
      </visual>
      <collision name="right_wheel_collision">
        <origin xyz="0 0 0" rpy="-3.141593 0 0"/>
        <geometry>
          <cylinder radius="0.2" length="0.1"/>
        </geometry>
      </collision>
    </link>

    <joint name="base_link" type="fixed">
      <origin xyz="0 0 0.4" rpy="0 0 0"/>
      <parent link="base_footprint"/>
      <child link="base_link"/>
    </joint>

    <joint name="left_wheel_joint" type="continuous">
      <origin xyz="0 0.45 -0.2" rpy="-1.570796 0 0"/>
      <parent link="base_link"/>
      <child link="left_wheel"/>
      <axis xyz="0 0 1.0"/>
      <limit effort="5000.0" velocity="60.0"/>
    </joint>

    <joint name="right_wheel_joint" type="continuous">
      <origin xyz="0 -0.45 -0.2" rpy="-1.570796 0 0"/>
      <parent link="base_link"/>
      <child link="right_wheel"/>
      <axis xyz="0 0 1.0"/>
      <limit effort="5000.0" velocity="60.0"/>
    </joint>

    <material name="Material.002">
      <color rgba="0 0 0.8 1.0"/>
    </material>

    <material name="Material.003">
      <color rgba="0.8 0 0 1.0"/>
    </material>

    <material name="Material.004">
      <color rgba="0.8 0 0 1.0"/>
    </material>

    <material name="Material.008">
      <color rgba="0.8 0 0 1.0"/>
    </material>

    <material name="Material.009">
      <color rgba="0 0 0.8 1.0"/>
    </material>

    <material name="Material.010">
      <color rgba="0.8 0 0 1.0"/>
    </material>

    <material name="Material.011">
      <color rgba="0 0.8 0 1.0"/>
    </material>

    <material name="Material.012">
      <color rgba="0 0.8 0 1.0"/>
    </material>

    <material name="Material.013">
      <color rgba="0 0.8 0 1.0"/>
    </material>

    <material name="Material.014">
      <color rgba="0 0.8 0 1.0"/>
    </material>

    <material name="Material.015">
      <color rgba="0.008067 0.1 0.012966 1.0"/>
    </material>

    <gazebo reference="base_link" >
      <mu1>0.0</mu1>
      <mu2>0.0</mu2>
    </gazebo>

    <gazebo reference="left_wheel" >
      <mu1>${wheel_friction}</mu1>
      <mu2>${wheel_friction}</mu2>
    </gazebo>

    <gazebo reference="right_wheel" >
      <mu1>${wheel_friction}</mu1>
      <mu2>${wheel_friction}</mu2>
    </gazebo>

  </robot>
